File: app/src/main/java/kr/gyeongmunbaek/imagesequenceanimaion/view/image/cache/CustomCache.java
package kr.gyeongmunbaek.imagesequenceanimaion.view.image.cache;

import android.content.Context;
import android.util.Log;

import java.lang.reflect.InvocationTargetException;
import java.util.HashMap;
import java.util.Map;

import kr.gyeongmunbaek.imagesequenceanimaion.view.image.BitmapCreatorParent;

/**
 * Created by gyeongmunbaek on 2017-10-26.
 */


public class CustomCache<T extends BitmapCreatorParent> {
    private int mCacheCount = 0;
    private Context mContext = null;
    private BitmapCreatorParent mRealClass = null;

    private Map<Integer, T> mMap = null;
    private int mLastIndex = 0;

    public CustomCache(Context pContext, int pCacheCount, BitmapCreatorParent pRealClass) {
        mCacheCount = pCacheCount;
        mContext = pContext;
        mRealClass = pRealClass;

        mMap = new HashMap<>();
    }

    public T get(Integer pKey) {
        T lTemp = mMap.get(pKey);

        if (lTemp == null) {
            int lIndex = pKey.intValue();
            int lPreIndex = (lIndex - mCacheCount);

            if (lPreIndex >= 0) {
                if (mMap.get(Integer.valueOf(lPreIndex)) != null) {
                    mMap.get(Integer.valueOf(lPreIndex)).recycle();
                    mMap.remove(Integer.valueOf(lPreIndex));
                }
            }
            lTemp = create(pKey);
            mMap.put(pKey, lTemp);
            mLastIndex = pKey.intValue();
        }

        return lTemp;
    }

    private T create(Integer pKey) {
        T lResultClass = null;
        try {
            Class<T> Clazz = (Class<T>) mRealClass.getClass();
            lResultClass = Clazz.getDeclaredConstructor(Context.class, Integer.class).newInstance(mContext, pKey);
        } catch (InstantiationException e) {
            e.printStackTrace();
            Log.i("GMBAEK", "InstantiationException : " + e.getMessage());
        } catch (IllegalAccessException e) {
            e.printStackTrace();
            Log.i("GMBAEK", "IllegalAccessException : " + e.getMessage());
        } catch (NoSuchMethodException e) {
            e.printStackTrace();
            Log.i("GMBAEK", "NoSuchMethodException : " + e.getMessage());
        } catch (InvocationTargetException e) {
            e.printStackTrace();
            Log.i("GMBAEK", "InvocationTargetException : " + e.getMessage());
        }

        return lResultClass;
    }

    public void evictAll() {
        for (int index = mLastIndex - mCacheCount; index <= mLastIndex; index++) {
            Integer lKey = Integer.valueOf(index);
            if (mMap.get(lKey) != null) {
                mMap.get(lKey).recycle();
                mMap.remove(lKey);
            }
        }
        mMap.clear();
        mMap = null;

        if (mRealClass != null) {
            mRealClass.recycle();
        }
    }

    public void evict(Integer pKey) {
        if (mMap.get(pKey) != null) {
            mMap.get(pKey).recycle();
            mMap.remove(pKey);
        }
    }

    public int size() {
        return mMap.size();
    }
}

File: app/src/main/java/kr/gyeongmunbaek/imagesequenceanimaion/view/image/cache/CustomCache.kt
package kr.gyeongmunbaek.imagesequenceanimaion.view.image.cache

import android.content.Context
import android.util.Log
import android.util.SparseArray

import java.lang.reflect.InvocationTargetException

import kr.gyeongmunbaek.imagesequenceanimaion.view.image.BitmapCreatorParent

/**
 * Created by gyeongmunbaek on 2017-10-26.
 */


class CustomCache<T : BitmapCreatorParent>(private val context: Context,
                                           private val cacheCount: Int,
                                           private val realClass: BitmapCreatorParent) {
    private val map: SparseArray<T> by lazy { SparseArray<T>() }
    private var lastIndex = 0

    operator fun get(key: Int): T? {
        var tempT = map[key]

        if (tempT == null) {
            val lPreIndex = key - cacheCount

            if (lPreIndex >= 0) {
                map[Integer.valueOf(lPreIndex)]?.let {
                    it.recycle()
                    map.remove(Integer.valueOf(lPreIndex))
                }
            }
            tempT = create(key)
            tempT?.let {
                map.put(key, it)
                lastIndex = key
            }
        }

        return tempT
    }

    private fun create(key: Int): T? {
        var lResultClass: T? = null
        try {
            val clazz = realClass.javaClass as Class<*>
            lResultClass = clazz.getDeclaredConstructor(Context::class.java, Int::class.java).newInstance(context, key) as T?
        } catch (e: InstantiationException) {
            e.printStackTrace()
            Log.i("GMBAEK", "InstantiationException : " + e.message)
        } catch (e: IllegalAccessException) {
            e.printStackTrace()
            Log.i("GMBAEK", "IllegalAccessException : " + e.message)
        } catch (e: NoSuchMethodException) {
            e.printStackTrace()
            Log.i("GMBAEK", "NoSuchMethodException : " + e.message)
        } catch (e: InvocationTargetException) {
            e.printStackTrace()
            Log.i("GMBAEK", "InvocationTargetException : " + e.message)
        }

        return lResultClass
    }

    fun evictAll() {
        for (index in lastIndex - cacheCount..lastIndex) {
            map[index]?.let {
                it.recycle()
                map.remove(index)
            }
        }
        map.clear()
        realClass.recycle()
    }

    fun evict(key: Int) {
        map[key]?.let {
           it.recycle()
            map.remove(key)
        }
    }

    fun size(): Int {
        return map.size()
    }
}
